### PR TITLE
*/Cargo.toml: Quaratine jemalloc with features

### PR DESCRIPTION
**Pull request by Hoverbear**
Signed-off-by: Ana Hobden <operator@hoverbear.org>

## What have you changed? (mandatory)

While I don't have any particular problem with jemalloc, I'd prefer it only be in the build graph when we want it to!

Currently, when you build TiKV with `--no-default-features --features "sse portable mimalloc"` (or `tcmalloc`) our friend `jemalloc` also comes into the build tree.

You can verify this with:

```bash
cargo build --no-default-features --features "portable mimalloc sse" --build-plan -Z unstable-options | grep jemalloc
```

You'll see `jemalloc` is part of the build plan regardless of feature.

This is problematic for targets like `musl` since `jemalloc` is not compatible... But having two allocators is generally a bad idea anyways.

This patch makes `jemalloc` only part of the build graph if it is enabled.

You can test by running:

```bash
cargo build --no-default-features --features "portable mimalloc sse" --build-plan -Z unstable-options | grep jemalloc
```

In this tree, and comparing it to the one on master. In this tree, the result is empty. In master, there is a result.

## What are the type of changes? (mandatory)

**Bugfix**, we already implemented this feature in other PRs, this one just fixes a couple missed flags.

## How has this PR been tested? (mandatory)

```bash
cargo build --no-default-features --features "portable mimalloc sse" --build-plan -Z unstable-options | grep jemalloc
```

## Does this PR affect documentation (docs) or release note? (mandatory)

No, it was found by following the documentation. :)

## Does this PR affect `tidb-ansible` update? (mandatory)

Nope

### CODE REVIEW

**Comment by brson** [comment]
Actually I don't see that either engine nor rust-rocksdb has default features, so I'm unsure what affect this PR is having.

**Comment by Hoverbear** [comment]
Whoops, those tikv_util changes are stray...

**Comment by breeswish** [comment]
Hmmm. So why not simply remove these default features of the malloc?

**Comment by siddontang** [comment]
As @breeswish said above, can we remove the default malloc feature? so we must pass allocator we want to use explicitly.

/cc @BusyJay

**Comment by Hoverbear** [comment]
@siddontang I think it is a big usability feature to support `cargo build` (run, fmt, test, bench....) without any extra options so I think it's important to keep *some* default in TiKV.

Perhaps it's better we adopt a strategy of not having default features on subcrates? So only TiKV has default features, and the subcrates must have features manually opted into?

**Comment by siddontang** [comment]
> So only TiKV has default features, and the subcrates must have features manually opted into

I think it is workable :-) 

/cc @BusyJay

**Comment by Hoverbear** [comment]
OK, I'll work on this

**Comment by Hoverbear** [comment]
Should be good to go, only crates with default features are cmd and tikv.

File: cmd/Cargo.toml [hunk 1]
```diff
@@ -49,7 +49,7 @@ hex = "0.3"
 vlog = "0.1.4"
 kvproto = { git = "https://github.com/pingcap/kvproto.git" }
 tikv_util = { path = "../components/tikv_util" }
-engine = { path = "../components/engine" }
+engine = { path = "../components/engine", default-features = false }
```
Comment (nrc) [inline cmd/Cargo.toml]: These changes to the dep on engine are not needed

Comment (brson) [reply cmd/Cargo.toml]: Likewise on tikv's engine dep - engine dosen't have any default features.

File: components/engine/Cargo.toml [hunk 2]
```diff
@@ -35,6 +35,7 @@ features = ["nightly"]
 [dependencies.engine_rocksdb]
 git = "https://github.com/pingcap/rust-rocksdb.git"
 package = "rocksdb"
+default-features = false
```
Comment (brson) [inline components/engine/Cargo.toml]: rut-rocksdb doesn't have any default features, at least on master.

**Comment by siddontang** [comment]
so is there a way to check that the crate has a default feature but we forget to forbid it?

File: cmd/Cargo.toml [hunk 3]
```diff
@@ -8,7 +8,7 @@ publish = false
 [features]
 default = ["tikv/default"]
 tcmalloc = ["tikv/tcmalloc"]
-jemalloc = ["tikv/jemalloc"]
+jemalloc = ["tikv/jemalloc", "engine/jemalloc"]
```
Comment (siddontang) [inline cmd/Cargo.toml]: em, why we need `engine/jemalloc` here?

File: Cargo.toml [hunk 4]
```diff
@@ -116,7 +116,7 @@ kvproto = { git = "https://github.com/pingcap/kvproto.git" }
 tikv_alloc = { path = "components/tikv_alloc", default-features = false }
 tikv_util = { path = "components/tikv_util" }
 log_wrappers = { path = "components/log_wrappers" }
-engine = { path = "components/engine" }
+engine = { path = "components/engine", default-features = false }
```
Comment (siddontang) [inline Cargo.toml]: why disable default for engine here?

**Comment by brson** [comment]
cc @Hoverbear

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hoverbear) [reply Cargo.toml]: I'm really sorry, I totally missed committing this. :(

Comment (Hoverbear) [reply Cargo.toml]: Because sometimes I do a terrible job with my `git add` globs. :(

**Pull request merged**